pkg/services/alerting/ticker.go

```diff
@@ -1,76 +1,64 @@
-// * it ticks on intervalSec marks or very shortly after. this provides a predictable load pattern
+// * it ticks on interval marks or very shortly after. this provides a predictable load pattern
 //   (this shouldn't cause too much load contention issues because the next steps in the pipeline just process at their own pace)
 // * the timestamps are used to mark "last datapoint to query for" and as such, are a configurable amount of seconds in the past
-// * because we want to allow:
-//   - a clean "resume where we left off" and "don't yield ticks we already did"
-//   - adjusting offset over time to compensate for storage backing up or getting fast and providing lower latency
-//   you specify a lastProcessed timestamp as well as an offset at creation, or runtime
 type Ticker struct {
 	C           chan time.Time
 	clock       clock.Clock
 	last        time.Time
-	offset      time.Duration
-	newOffset   chan time.Duration
-	intervalSec int64
-	paused      bool
+	interval    time.Duration
+	behindTicks prometheus.Gauge
 }
 
-// NewTicker returns a ticker that ticks on intervalSec marks or very shortly after, and never drops ticks
-func NewTicker(last time.Time, initialOffset time.Duration, c clock.Clock, intervalSec int64) *Ticker {
+// NewTicker returns a ticker that ticks on interval marks or very shortly after, and never drops ticks. interval should not be negative or zero
+func NewTicker(c clock.Clock, interval time.Duration) *Ticker {
+	if interval <= 0 {
+		panic(fmt.Errorf("non-positive interval [%v] is not allowed", interval))
+	}
+
 	t := &Ticker{
-		C:           make(chan time.Time),
-		clock:       c,
-		last:        last,
-		offset:      initialOffset,
-		newOffset:   make(chan time.Duration),
-		intervalSec: intervalSec,
+		C:        make(chan time.Time),
+		clock:    c,
+		last:     c.Now(),
+		interval: interval,
+		behindTicks: promauto.With(prometheus.DefaultRegisterer).NewGauge(prometheus.GaugeOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "ticker_behind_ticks",
+			Help:      "The number of ticks the ticker is lagging behind",
+		}),
 	}
 	go t.run()
 	return t
 }
 
 func (t *Ticker) run() {
 	for {
-		next := t.last.Add(time.Duration(t.intervalSec) * time.Second)
-		diff := t.clock.Now().Add(-t.offset).Sub(next)
+		next := t.last.Add(t.interval)  // calculate the time of the next tick
+		diff := t.clock.Now().Sub(next) // calculate the difference between the current time and the next tick
+		// if difference is not negative, then it should tick
 		if diff >= 0 {
-			if !t.paused {
-				t.C <- next
-			}
+			t.C <- next
+			// update metric after the push to the channel to guarantee that the tick is delivered as soon as possible
+			t.behindTicks.Set(float64(diff.Nanoseconds() / t.interval.Nanoseconds()))
 			t.last = next
 			continue
 		}
 		// tick is too young. try again when ...
 		select {
 		case <-t.clock.After(-diff): // ...it'll definitely be old enough
```

Comment:
Could we simplify this a bit further by dropping the `select` entirely and just blocking on the `After` call? This has been reduced to just a single code path.